feat(s2n-quic): add jumbo frame support (#1630)
aws/s2n-quic: examples/jumbo-frame/README.md

# Jumbo Frame Support
The standard ethernet Maximum Transmission Unit - MTU - is 1500 bytes. Jumbo frames are ethernet frames that support more than 1500 bytes. Jumbo frames will often support approximately 9000 bytes, although this is an implementation specific detail and you should check to see what your network supports.

## why use jumbo frames

> Some overheads occur per-packet. For example, udp packet headers are included on each packet. Using jumbo frames decreases the overhead / payload ratio, so communication is more efficienct.

Reviewer: ```suggestion
Some overheads occur per-packet. For example, UDP packet headers are included on each packet. Using jumbo frames decreases the overhead / payload ratio, so communication is more efficienct.
```

CPU utilization also benefits from jumbo frame usage.

## how to use jumbo frames
Jumbo frames can only be used if they are supported by the network. AWS supports jumbo frames in the specific circumstances listed [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/network_mtu.html)

The `tracepath` utility can be used to check if jumbo frames are supported on a local machine.

```console
[ec2-user@ip-1-1-1-1 jumbo-frame]$ tracepath localhost
 1:  localhost                                             0.035ms reached
     Resume: pmtu 65535 hops 1 back 1
```

> The `pmtu 65535` indicates that `localhost` has a maximum tranmission unit of 65,535 bytes. Networks are unlikely to support an mtu this large, but this confirms that our demo with a 9,001 byte mtu will work properly.

Reviewer: ```suggestion
The `pmtu 65535` indicates that `localhost` has a maximum tranmission unit of 65,535 bytes. Networks are unlikely to support an MTU this large, but this confirms that our demo with a 9,001 byte MTU will work properly.
```

   Compiling rustls-provider v0.1.0 (/home/ec2-user/workspace/s2n-quic/examples/jumbo-frame)
    Finished dev [unoptimized + debuginfo] target(s) in 3.23s
     Running `target/debug/jumbo_server`
Listening for a connection

```

In the other terminal, launch the client.
```console
[ec2-user@ip-172-31-28-149 jumbo-frame]$ cargo run --bin jumbo_client
   Compiling rustls-provider v0.1.0 (/home/ec2-user/workspace/s2n-quic/examples/jumbo-frame)
    Finished dev [unoptimized + debuginfo] target(s) in 3.44s
     Running `target/debug/jumbo_client`
MtuUpdated { path_id: 0, mtu: 1200, cause: NewPath }
MtuUpdated { path_id: 0, mtu: 1472, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 5222, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 7097, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8035, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8504, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8738, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8855, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8914, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8943, cause: ProbeAcknowledged }
```

> The output confirms that larged MTU's (jumbo frames) have been probed.

Reviewer: ```suggestion
The output confirms that large MTU's (jumbo frames) have been probed.
```

> Note that both endpoints, client and server, must enable jumbo frames for either side to use them. If a client with `max_mtu=9001` tries to probe a server with `max_mtu=1500`, then the server will drop the probes larger than 1500 bytes, and the client will continue to use the 1472 byte mtu that it initially negotiated.

Reviewer: ```suggestion
Note that both endpoints, client and server, must enable jumbo frames for either side to use them. If a client with `max_mtu=9001` tries to probe a server with `max_mtu=1500`, then the server will drop the probes larger than 1500 bytes, and the client will continue to use the 1472 byte MTU that it initially negotiated.
```

## ensure jumbo frames are being used
The s2n-quic events system has an event for MTU updates. The subscriber defined in `lib.rs` captures this event to output messages like those below.
```
MtuUpdated { path_id: 0, mtu: 8943, cause: ProbeAcknowledged }
```

> This event can be used to verify that jumbo frames are being used. Alternatey, tools like `tcpdump` with a packet analysis tool like `wireshark` can confirm that jumbo packets are being sent and recieved.

Reviewer: ```suggestion
This event can be used to verify that jumbo frames are being used. Alternatively, tools like `tcpdump` with a packet analysis tool like `wireshark` can confirm that jumbo packets are being sent and received.
```

## probing strategy

> Quic implements Datagram Packetization Layer Path Maximum Transmission Unit Discovery, or DPLPMTUD for short. This is described in [RFC8899](https://www.rfc-editor.org/info/rfc8899). The strategy is as follows. To determine if an `X` byte MTU is supported, send a packet of `X` bytes.

Reviewer: ```suggestion
s2n-quic implements Datagram Packetization Layer Path Maximum Transmission Unit Discovery, or DPLPMTUD for short. This is described in [RFC8899](https://www.rfc-editor.org/info/rfc8899). The strategy is as follows. To determine if an `X` byte MTU is supported, send a packet of `X` bytes.
```

> # Jumbo Frame Support

Author: Had some offline discussion that the language is a bit too informal in this readme for public facing AWS stuff, updating in next commit to remove first and second person references, exclamation points, and most adjectives/adverbs.

[ec2-user@ip-172-31-28-149 jumbo-frame]$ cargo run --bin jumbo_server
   Compiling rustls-provider v0.1.0 (/home/ec2-user/workspace/s2n-quic/examples/jumbo-frame)
    Finished dev [unoptimized + debuginfo] target(s) in 3.23s
     Running `target/debug/jumbo_server`
Listening for a connection

```

In the other terminal, go ahead and start the client
```console
[ec2-user@ip-172-31-28-149 jumbo-frame]$ cargo run --bin jumbo_client
   Compiling rustls-provider v0.1.0 (/home/ec2-user/workspace/s2n-quic/examples/jumbo-frame)
    Finished dev [unoptimized + debuginfo] target(s) in 3.44s
     Running `target/debug/jumbo_client`
MtuUpdated { path_id: 0, mtu: 1200, cause: NewPath }
MtuUpdated { path_id: 0, mtu: 1472, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 5222, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 7097, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8035, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8504, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8738, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8855, cause: ProbeAcknowledged }
MtuUpdated { path_id: 0, mtu: 8914, cause: ProbeAcknowledged }

> MtuUpdated { path_id: 0, mtu: 8943, cause: ProbeAcknowledged }

Reviewer 2: Not blocking this PR but I'm wondering if this is the best strategy... That's quite a few probes that we're sending. For networks that you know do support jumbo frames, like EC2, it might be more optimal to try and jump to the highest packet size.

Author: Attempting to summarize offline conversation here.

Starting the probe at the lower end and going upwards was chosen to keep the probing overhead low on networks when
- users were sending small packet &&
- users enabled jumbo frames, but they were not supported on the network

That case performs particularly poorly because
- For each acked packet, the congestion window grows by packet_size
- so the loss of a 9,000 byte probe slows down the growth of the congestion window

That being said, user's that are sure of their network topology and MTU would gain some benefit by immediately probing for the maximum mtu, so we should figure out a way to unblock them, although that work is outside the scope of this PR.
